drivers/io/io_fip.c

```diff
@@ -32,11 +36,22 @@ typedef struct {
 	fip_toc_entry_t entry;
 } file_state_t;
 
+/* Maintain backend handles and file state per FIP device */
+typedef struct {
+	uintptr_t dev_spec;
+	uintptr_t backend_dev_handle;
+	uintptr_t backend_image_spec;
+	/* One file per device is maintained */
+	file_state_t current_file;
```

Comment:
@ruchi393, Thanks for the rework. In the current io_memmap driver, I realized that it only allows a single file to be open.  Having several open FIP files at the same time, implies the same from io_memmap driver. Hence supporting multiple files open at the io_fip layer is not very useful without corresponding io_memmap support.

From your description of the ARM-Software/tf-issues#591, You seem to require multiple FIP devices but not necessarily several files open at the same time.  Since the generic load_image() always loads the image in full before moving to the next image (i.e only a single file is `active` at a given time), multiple fip devices can be supported without having multiple file support.  Perhaps the structure can be simplified as : 

```
typedef struct {
	uintptr_t dev_spec;
} fip_dev_state_t;
```
Only a single global file state `current_file` would be maintained which implies that the file has to be opened, read and closed before the next file is accessed (which may be in another FIP device). Would that work for you ?

Reply:
You are right about the requirement.  At a time I will have only one file open of a single device.  I will explore and see what changes are required to maintain a global file for each device.

Reply:
@soby-mathew , I checked the memmap code in io_memmap.c . There I see that in memmap_block_open function, there is a check on if file is in use and this gets cleared in memmap_block_close function. In fip driver, the backend's open and close functions are not spread across functions.   In BL2 we are executing on a single core. IIUC, this would mean that even though backend driver has a single file, it shouldn't matter. Please help correct me in case I am missing out on something.

Reply:
@ruchi393, Your understanding is correct. TBH, my understanding of the I/O layer is not strong. 

> Block devices seem to have multiple block devices support in it. Would it be better to expand the io_memmap driver to add multiple file support. What do you think ?

AIU, the layering of the I/O layer when accessed for FIP is as below:
```
       io_storage
           |
           v
        io_fip
           |
      -------------
     |             |
io_memmap     io_block
```
The `io_block` layer can be used to access the block devices in addition to via FIP. For the `io_memmap`  they are only used via FIP (the IO memmap drivers are not used for direct access).
This is why I think why io_block has multiple file capability whereas io_memmap didn't.

Preferably we don't want to add more support in IO layer unless needed.

> In fip driver, the backend's open and close functions are not spread across functions. In BL2 we are executing on a single core. IIUC, this would mean that even though backend driver has a single file, it shouldn't matter

Yes, the io_fip driver opens and closes the underlaying io_layer for all io operation. I think having a single backend handle should be alright.

> Basically what you are suggesting doesn't require any change in FIP driver. dev_spec is a dummy entity , not used after device_open function

When you want to open more than 1 FIP device at a time (i.e. multiple calls to fip_dev_open()), you would need to support returning multiple handles. Since internally the io_fip driver can only support one open file at time, yes, the dev_spec would remain a dummy entity not used after fip_dev_open().

Reply:
> Do you see any issue with multiple backend handles of memmap driver ? I thought it made the design clearer.

> With above design I thought as per your earlier suggestion having multiple backend driver made the flow very clean for the multiple FIP support.

I initially misunderstood that you needed multiple file support in io_fip driver which is not the case. Multiple open file support in io_fip mandates the same support from the all the backends and this brings in additional memory overhead and complexity which is not required now.  Hence would prefer to keep it simple and just allow muiple fip devices for now.

> Please suggest if you want me to remove the multiple backend support and revert to the first patch which I had sent.

The patch would be similar to the first version except that you don't need to add `current_file` field to  `fip_dev_state_t`. Just have the `dev_spec` as the only field. Possibly suitable comments can be added to indicate what is involved if mutiple open files have to be supported and can be taken up when there is a requirement.

Reply:
I have updated the pull request with the requested changes. Please review.